File: mobile-app/lib/core/utils/course_access.dart
import '../../models/course.dart';
import '../../models/lesson.dart';
/// Enforces 1 free demo video for paid courses until the user has purchased.
class CourseAccess {
  CourseAccess._();

  /// True when the user bought or subscribed to this course (not preview-only).
  static bool isCoursePurchased(
    Course course, {
    bool subscriptionActive = false,
  }) {
    if (!course.isPaid) return true;
    return course.hasPurchased || subscriptionActive;
  }

  static bool hasFullPlaybackAccess(
    Course course, {
    bool subscriptionActive = false,
  }) {
    if (!course.isPaid) return true;
    if (course.hasAccess || course.hasPurchased || subscriptionActive) {
      return true;
    }
    return false;
  }

  static Course applyPlaybackLocks(
    Course course, {
    bool subscriptionActive = false,
  }) {
    if (hasFullPlaybackAccess(course, subscriptionActive: subscriptionActive)) {
      return _unlockAllVideos(
        course,
        subscriptionActive: subscriptionActive,
      );
    }

    var previewGranted = false;
    final lessons = course.lessons.map((lesson) {
      if (lesson.videos.any((video) => video.isLocked)) {
        return lesson;
      }

      final videos = lesson.videos.map((video) {
        if (!previewGranted && video.hasPlayableSource) {
          previewGranted = true;
          return video.copyWith(isLocked: false);
        }
        return video.copyWith(
          isLocked: true,
          previewVideoUrl: video.previewVideoUrl ??
              (video.videoUrl.isNotEmpty
                  ? video.videoUrl
                  : video.hlsUrl),
          videoUrl: '',
        );
      }).toList();

      return Lesson(
        id: lesson.id,
        title: lesson.title,
        order: lesson.order,
        description: lesson.description,
        isFree: lesson.isFree,
        isPublished: lesson.isPublished,
        videos: videos,
      );
    }).toList();

    return Course(
      id: course.id,
      title: course.title,
      slug: course.slug,
      description: course.description,
      thumbnail: course.thumbnail,
      previewVideoUrl: course.previewVideoUrl,
      categoryId: course.categoryId,
      categoryName: course.categoryName,
      level: course.level,
      isPublished: course.isPublished,
      pricing: course.pricing,
      instructorName: course.instructorName,
      lessons: lessons,
      videoCount: lessons.fold<int>(0, (sum, l) => sum + l.videoCount),
      isPaid: course.isPaid,
      hasAccess: course.hasAccess,
      hasPurchased: course.hasPurchased,
      previewVideoCount: 1,
    );
  }

  static Course _unlockAllVideos(
    Course course, {
    bool subscriptionActive = false,
  }) {
    final lessons = course.lessons.map((lesson) {
      final videos = lesson.videos.map((video) {
        final playbackUrl = video.videoUrl.isNotEmpty
            ? video.videoUrl
            : (video.previewVideoUrl ?? '');
        return video.copyWith(
          isLocked: false,
          videoUrl: playbackUrl,
        );
      }).toList();

      return Lesson(
        id: lesson.id,
        title: lesson.title,
        order: lesson.order,
        description: lesson.description,
        isFree: lesson.isFree,
        isPublished: lesson.isPublished,
        videos: videos,
      );
    }).toList();

    final playableCount =
        lessons.fold<int>(0, (sum, lesson) => sum + lesson.videos.length);

    return Course(
      id: course.id,
      title: course.title,
      slug: course.slug,
      description: course.description,
      thumbnail: course.thumbnail,
      previewVideoUrl: course.previewVideoUrl,
      categoryId: course.categoryId,
      categoryName: course.categoryName,
      level: course.level,
      isPublished: course.isPublished,
      pricing: course.pricing,
      instructorName: course.instructorName,
      lessons: lessons,
      videoCount: playableCount > 0 ? playableCount : course.videoCount,
      isPaid: course.isPaid,
      hasAccess: true,
      hasPurchased: course.hasPurchased || subscriptionActive,
      previewVideoCount: course.previewVideoCount,
      hasPlayableVideos: course.hasPlayableVideos,
    );
  }

  static String? firstPlayableVideoId(Course course) {
    for (final lesson in course.lessons) {
      for (final video in lesson.videos) {
        if (!video.isLocked && video.hasPlayableSource) return video.id;
      }
    }
    return null;
  }

  static bool isVideoPlayable(
    Course course,
    String videoId, {
    bool subscriptionActive = false,
  }) {
    final locked = applyPlaybackLocks(
      course,
      subscriptionActive: subscriptionActive,
    );
    for (final lesson in locked.lessons) {
      for (final video in lesson.videos) {
        if (video.id == videoId) {
          return !video.isLocked && video.hasPlayableSource;
        }
      }
    }
    return false;
  }
}

File: backend/utils/courseAccess.js
export const isPaidCourse = (course) => {
  const pricing = course?.pricing;
  if (!pricing) return false;
  return (
    (pricing.monthly ?? 0) > 0 ||
    (pricing.yearly ?? 0) > 0 ||
    (pricing.lifetime ?? 0) > 0
  );
};

const hasActiveSubscription = (user, courseId) => {
  if (!user || !courseId) return false;

  const targetId = courseId.toString();
  return user.subscriptions?.some((sub) => {
    if (sub.status !== "active") return false;
    if (sub.endDate && new Date(sub.endDate) <= new Date()) return false;
    const subCourseId = sub.course?._id?.toString() ?? sub.course?.toString();
    return subCourseId === targetId;
  });
};

/** True only when the user bought/enrolled in this exact course. */
export const userHasPurchasedCourse = (user, courseId) =>
  hasActiveSubscription(user, courseId);

/** Playback access (includes admin/instructor bypass). */
export const userHasCourseAccess = (user, courseId) => {
  if (!user || !courseId) return false;
  if (["admin", "instructor"].includes(user.role)) return true;
  return hasActiveSubscription(user, courseId);
};

import { pickVideoUrl, pickVideoThumbnail } from "./courseThumbnail.js";

const stripLockedVideoMedia = (video) => {
  const previewVideoUrl = pickVideoUrl(video);
  const thumbnail = pickVideoThumbnail(video) || video?.thumbnail;

  const sanitized = {
    ...video,
    isLocked: true,
    thumbnail: thumbnail || video?.thumbnail,
    previewVideoUrl,
  };
  delete sanitized.videoUrl;
  delete sanitized.videoKey;
  delete sanitized.externalUrl;
  return sanitized;
};

/**
 * Every paid course shows 1 free preview video until purchased.
 * Only the purchased course is fully unlocked — other courses stay locked.
 */
export const applyCourseAccess = (courseData, user) => {
  const paid = isPaidCourse(courseData);
  const courseId = courseData._id?.toString();
  const hasPurchased = userHasPurchasedCourse(user, courseId);
  const hasAccess = !paid ? true : userHasCourseAccess(user, courseId);

  courseData.isPaid = paid;
  courseData.hasPurchased = hasPurchased;
  courseData.hasAccess = hasAccess;
  courseData.previewVideoCount = 1;

  if (hasAccess) {
    for (const lesson of courseData.lessons || []) {
      lesson.videos = (lesson.videos || []).map((video) => ({
        ...video,
        isLocked: false,
      }));
    }
    return courseData;
  }

  let previewGranted = false;

  for (const lesson of courseData.lessons || []) {
    lesson.videos = (lesson.videos || []).map((video) => {
      if (!previewGranted) {
        previewGranted = true;
        return { ...video, isLocked: false };
      }
      return stripLockedVideoMedia(video);
    });
  }

  return courseData;
};
